packages/node/src/retryHandler.ts

```diff
@@ -0,0 +1,186 @@
+      if (this._shouldRetryEvents()) {
+        this._onEventsError(events);
+      }
+    } finally {
+      return response;
+    }
+  }
+
+  private _setupDefaultTransport(): Transport {
+    const transportOptions: TransportOptions = {
+      serverUrl: this._options.serverUrl,
+      headers: {
+        'Content-Type': 'application/json',
+      },
+    };
+    return new HTTPTransport(transportOptions);
+  }
+
+  private _shouldRetryEvents(): boolean {
+    if (typeof this._options.maxRetries !== 'number' || this._options.maxRetries <= 0) {
+      return false;
+    }
+
+    const bufferLimit = this._options.maxCachedEvents;
+
+    return this._eventsInRetry < bufferLimit;
+  }
+
+  // Sends events with ids currently in active retry buffers straight
+  // to the retry buffer they should be in
+  private _pruneEvents(events: ReadonlyArray<Event>): Array<Event> {
+    const prunedEvents: Array<Event> = [];
+    events.forEach(event => {
+      const { user_id: userId = '', device_id: deviceId = '' } = event;
+      if (userId || deviceId) {
+        const retryBuffer = this._getRetryBuffer(userId, deviceId);
+        if (retryBuffer?.length) {
+          retryBuffer.push(event);
+          this._eventsInRetry++;
+        } else {
+          prunedEvents.push(event);
+        }
+      }
+    });
+
+    return prunedEvents;
+  }
+
+  private _getPayload(events: ReadonlyArray<Event>): Payload {
+    return {
+      api_key: this._apiKey,
+      events,
+    };
+  }
+
+  private _getRetryBuffer(userId: string, deviceId: string): Array<Event> | null {
+    const deviceToBufferMap = this._idToBuffer.get(userId);
+    if (!deviceToBufferMap) {
+      return null;
+    }
+
+    return deviceToBufferMap.get(deviceId) || null;
+  }
+
+  // cleans up the id to buffer map if the job is done
+  private _cleanUpBuffer(userId: string, deviceId: string): void {
+    const deviceToBufferMap = this._idToBuffer.get(userId);
+    if (!deviceToBufferMap) {
+      return;
+    }
+
+    const eventsToRetry = deviceToBufferMap.get(deviceId);
+    if (!eventsToRetry?.length) {
+      deviceToBufferMap.delete(deviceId);
+    }
+
+    if (deviceToBufferMap.size === 0) {
+      this._idToBuffer.delete(userId);
+    }
+  }
+
+  private _onEventsError(events: ReadonlyArray<Event>): void {
+    events.forEach((event: Event) => {
+      const { user_id: userId = '', device_id: deviceId = '' } = event;
+      if (userId || deviceId) {
+        let deviceToBufferMap = this._idToBuffer.get(userId);
+        if (!deviceToBufferMap) {
+          deviceToBufferMap = new Map<string, Array<Event>>();
+          this._idToBuffer.set(userId, deviceToBufferMap);
+        }
+
+        let retryBuffer = deviceToBufferMap.get(deviceId);
+        if (!retryBuffer) {
+          retryBuffer = [];
+          deviceToBufferMap.set(deviceId, retryBuffer);
+          // In the next event loop, start retrying these events
+          setImmediate(() => this._retryEventsOnLoop(userId, deviceId));
```

Comment:
I think this is still not firing on the next event loop, though it's definitely better than `nextTick`

https://nodejs.org/en/docs/guides/event-loop-timers-and-nexttick/